src/ChannelBase.php

```diff
@@ -47,4 +47,84 @@ public static function optionsWidget(array &$element, array $options) {
   public function enabled() {
     return !empty($this->options['toggle']);
   }
+
+  /**
+   * Fetches the node object we want to share.
+   *
+   * In case of `campaigninon_thankyou_page`s, the parent node
+   * will be loaded (e.g. the corresponding petition).
+   *
+   * @return object
+   *   An array containing the node object and the URL query.
+   */
+  protected function getNode() {
+    $node = $this->block->getNode();
+    $link = $this->block->getLink();
+    $query['path'] = $link['path'];
+    if (isset($link['query'])) {
+      $query['query'] = $link['query'];
+    }
+    if ($p_node = menu_get_object('node', 1, $query['path'])) {
+      $node = $p_node;
+    }
+    if ($node && $query['path'] == 'node/' . $node->nid) {
+      unset($query['path']);
+    }
+
+    return array(
+      'node' => $node,
+      'query' => $query,
+    );
+  }
+
+  /**
+   * Generates the URL to be shared.
+   *
+   * @param string $utm_source
+   *   The `utm_source` query parameter's value (e.g. `twitter_share`).
+   * @param object|null $node
+   *   The `node` object containing the `node`'s `nid`. If `NULL` is
+   *   provided, the node will be fetched via `getNode()`.
+   *
+   * @return string|null
+   *   The generated URL or NULL if no node could be fetched.
```

Comment:
I think this should also work if no node is present (although we base a lot of special behavior on nodes). If this always returns properly all the checks in callers are not needed anymore.

Reply:
Which value should be returned if it's not possible to retrieve a node? `$this->block->getUrl()`?